tensorflow/core/kernels/BUILD

```diff
@@ -4029,7 +4029,7 @@ tf_kernel_library(
 tf_kernel_library(
     name = "bias_op",
     prefix = "bias_op",
-    deps = NN_DEPS + [":redux_functor"] + if_cuda([
+    deps = NN_DEPS + [":redux_functor"] + if_cuda_is_configured([
```

Comment:
This does not work internally.
I am not sure what the difference between `if_cuda` and `if_cuda_is_configured` is.
But to be able to merge this, someone will need to take a look into this and maybe even change the macros internally.
CC @chsigg @ezhulenev @jlebar

Reply:
hi @gunan,

If possible, can you please shed some light on the nature of the problems introduced by this change. We would like to understand the problem(s) associated with this change, to see if there is anything we can do on our end to avoid/workaround them.

thanks

deven

Reply:
Internally, there is no configure script. So `if_cuda_is_configured` semantically does not make sense for us internally.

Also, I am trying to understand what the differences between `if_cuda` and `if_cuda_is_configured` are. We are trying to move to a state where we do not have to run configure script, so only `if_cuda` will make sense in such a case.

Reply:
See the discussion between @chsigg and myself here for some info on the differences between `if_cuda` and `if_cuda_is_configured` : https://github.com/tensorflow/tensorflow/pull/26753#discussion_r267022934

>Internally, there is no configure script.

Do you still need to specify the env var `TF_NEED_CUDA=1` to build TF with CUDA support? 
* If that is the case then `if_cuda_is_configured` should still work.  
* Else if there is some different mechanism to specify building TF with CUDA support, then we should update this routine (https://github.com/tensorflow/tensorflow/blob/master/third_party/gpus/cuda_configure.bzl#L337) to trigger using that mechanism in addition to the env var. That should make the `if_cuda_is_configured` work without running the configure script.

@chsigg please confirm the accuracy of above. thanks.

Reply:
>  Do you still need to specify the env var TF_NEED_CUDA=1 to build TF with CUDA support?
No, we do not. Even in opensource, we would like to get to a state where all you need to do is `bazel build --config=cuda //tensorflow/...`

Internally, and soon externally, all you need to do for building with CUDA support is add `--config=cuda` flag. That is all that is available to bazel to decide what to use when building.

Reply:
In that case, how does routine I referred to above work (internally)? if it does not use the env var, it must use some other piece of information that is passed down to it by bazel.

Reply:
Here are the main things that happen when we do `--config=cuda`
https://github.com/tensorflow/tensorflow/blob/master/.bazelrc#L48

Internally, we have different defines, but it uses bazel's defines.

Reply:
hmm...I think `if_cuda_is_configured` will/should continue to work as long as `enable_cuda` routine (https://github.com/tensorflow/tensorflow/blob/master/third_party/gpus/cuda_configure.bzl#L337) returns `True` when TF ib being built with CUDA.

In the public repo, the `enable_cuda` implementation uses the `TF_NEED_CUDA` env var (which is set in the configure script), but my guess is that it uses something else internally at G. But as long as that routine returns the correct value, the `if_cuda_is_configured` routine will also work correctly (maybe we need to change its name to `if_cuda_is_enabled`).  Is this assumption correct? Irrespective, what is the path forward for resolving this PR?

thanks again.

Reply:
Well, externally, sure it can work. But any code that touches our core files, we cannot import `if_cuda_is_configured`. So for the purpose of this PR, either someone internal will need to take a look into this to make it work, or we need to find a way to make your code work with `if_cuda`. Unfortunately, I do not have cycles to investigate this.

In my opinion, it does not matter if cuda is configured or not. Whatever you set during configure, is irrelevant. You either build with cuda, or you do not. Even if you configure with both CUDA and ROCM support, unless you add `--config=rocm` or `--config=cuda` what you build should be equivalent to the CPU builds, as if you ran configure and did not set CUDA or ROCM.

Reply:
>So for the purpose of this PR, either someone internal will need to take a look into this to make it work, or we need to find a way to make your code work with if_cuda. Unfortunately, I do not have cycles to investigate this.

Due to the nature of the error, it does not seem that this is an issue we can solve by changing things exclusively on the ROCm side of the fence..we are going to need help. Given that, and also that this is (one of the) blocking issues preventing us from filing further PRs (for upstreaming ROCm support), we need to resolve this PR ASAP. What are our options towards that end?

@chsigg, your suggestion of `if_gpu(cuda_arg, rocm_arg, defaut_arg)` would be one potential solution here. How would we go about implementing + testing it out? 

thanks again

Reply:
I think it makes most sense if we clean this up from our end. I agree with Gunhan that if_cuda_is_configured (and .._compat) should go away. As far as I understand that macro was introduced to allow linking against CUDA libraries (say, cuDNN) while not invoking nvcc to build CUDA code. I'm not sure this is useful, but probably it won't be trivial to unify them again more than two years later. I will take a look and report back.

Reply:
@chsigg @gunan, let me know if / how I can help develop+test the new solution, will be more than happy to help out.

In the meantime, i will push out a commit soon, that works around this issue by introducing a new `if_cuda_or_rocm` function. This function works the same as `if_cuda` / `if_rocm`, i.e. it will return the `if_true` arg when either CUDA or ROCM is enabled. Using this function to specify the common dependencies, works around the bazel error. However, this solution is less than ideal because it needs to duplicate the logic within `if_cuda` / `if_rocm`.  Please review the commit and let me know if this is acceptable as a temporary workaround, while we work on the proper fix.

Reply:
This looks good as a temporary solution, thanks. I will run our internal tests and try to merge it.

Reply:
@chsigg thank you! 

I have pushed out a rebase to keep just the last commit and sync up to the tip.